README
A hoping-to-become-professional-grade web framework in Julia - at early stage, a lot more left to do, not ready for prime time yet. 

Done so far: Database low level abstraction layer on top of Postgres, ORM, Migrations, Tasks runner, Tests runner, Logging, Controllers, extended routing based on Mux
Left to do: the view layer, asset management, refactor router, maybe add support for working directly with HttpServer

CHANGELOG.md
# Changelog

## v1.9.1 - 2020-12-02

* generation of testable Genie apps
* better loading of secrets
* bug fixes

## v1.7.4 - 2020-10-22

* removed broken hooks from web sockets routing
* better exception handling in FileSessions.jl
* Secure => true sessions flag

## v1.7 - 2020-10-10

* reexport Router

## v1.3 - 2020-09-29

* support for running the HTTP/S and WS/S on the same port
* Renderer.Http does not output attributes with value `nothing`
* for new apps `config.server_handle_static_files` is enabled by default in production but warns (previously false)
* removed outdated asset pipeline files from `files/`

## v1.2 - 2020-07-21

* added Requests.getheaders
* added Requests.getrequest
* deps update
* removed Revise from where not used

## v1 - 2020-07-21

* added support for SSL to Genie apps
* added SSL initializer
* bundled localhost cert and key
* automatically forwarding all extra arguments from `Genie.up` to `HTTP.serve` `HTTP.listen`
* added tests
* fixed failing tests
* improved content negotiation and default error responses

## v0.32.0 - 2020-07-02

* new html() method, combining string view and file layout
* removed forgotten debug @show from Plugins
* fixed issue with <template> tags being stripped away
* fixed issue with missing <!DOCTYPE>
* deps update
* `Router.get_route` warns by default when named route not found (previously errored)

## v0.31.5 - 2020-07-01

* fixed issue with emtpy SECRET_TOKEN
* Session improvements
* more tests
* Markdown views fixes
* fixed `Genie.serve` for serving static websites

## v0.31.4 - 2020-07-01

* fixed issue with cookie encryption/decryption
* support for "inline" markdown views
* more tests

## v0.31.3 - 2020-06-30

* fix issue with CORS "*"
* documentation improvements
* doc comments

## v0.21.0 - 2020-04-23

* names of apps and resources are automatically cleaned (illegal Julia vars chars removed)
* support for file uploads with UTF-8 chars
* deps updates
* docs updates

## v0.25.2 - 2020-02-03

* bug fixes
* performance improvements
* deps update

## v0.22.9 - 2020-01-06

* refactoring of the Rendering engine (**breaking**)
* deps update

## v0.22.9 - 2019-12-24

* enhanced support for WebChannels and websockets operations

## v0.22.8 - 2019-12-05

* more tests
* support for `if` blocks
* internal API refactoring

## v0.22.7 - 2019-12-04

* added support for `if` blocks
* performance optimisations

## v0.22.6 - 2019-12-03

* bug fixes
* deps update

## v0.22.5 - 2019-11-30

* added type hints to renderer's `invokelastest` calls

## v0.22.4 - 2019-11-30

* fixed issue with JSON rendering where view variables won't get updated

## v0.22.3 - 2019-11-15

* logic to normalize the case of request headers to "Request-Header-Key"

## v0.21 - 2019-11-15

* deps update
* `Router.@params` is now set with default values
* vars in view can now be accessed directly (without `@vars`)
* internal API enhancements and bug fixes

## v0.20.0 - 2019-11-10

* internal rewrite and API cleanup (**breaking**)
* removal of Nullables and YAML deps
* deps update
* over 50 new unit tests added

## v0.19.1 - 2019-10-18

* Documentation improvements for `Genie.Deploy`
* Dockerfile generation enhancements
* Deps updates

## v0.19.0 - 2019-09-25

* New API for deployments
* Docker integration: generators for Dockerfile and API for building and running containers
* Heroku integration: API for deploying dockerized Genie apps to Heroku
* Extended API for app generation: `newapp_webservice`, `newapp_mvc`, and `newapp_fullstack`
* Deps updates

## v0.18.1 - 2019-09-25

* fixes an issue with parsing JSON payloads
* deps updates

## v0.18.0 - 2019-09-20

* pluralized some of the folders for consistent naming: sessions, tasks, and tests
* fixed an issue with generating a new task
* deps updates

## v0.17.1 - 2019-09-09

* improved support for testing
* deps updates
* small bug fixes

## v0.17.0 - 2019-08-31

* fully migrated logging to Julia's native logger (**breaking**)
* reintroduced support for logging to file via `LoggingExtras` (new app dependency)
* added a new command `bin/serverinteractive` allowing to start the web server interactively
* fixed issue with webserver port env settings being overwritten by default settings
* small changes to better support logging
* moved welcome info out of app files into core files
* added default `favicon.ico` file to avoid annoying 404 errors from automatic browsers requests
* cleaned up the env files
* small documentation updates
* dependencies updates

## v0.16.0 - 2019-08-29

* switched to native Julia logging (automatic logging to file for now is disabled, will come back in a future version)
* the `log` function has been removed (**breaking**)
* added support for embedded Julia within HTML arguments
* cleaned up HTML rendering
* refactored cache adapters loading to be less hacky and more performant
* refactored session adapters loading to be less hacky and more performant
* consolidated the `Helpers` API into the `Requests` and `Sessions` modules and removed `Helpers` module (**breaking**)
* added new `Exceptions` module defining the `ExceptionalResponse` type
* added extra `@params` pointing to the currently matched route and webchannel
* fixed broken `Cookies` and `Session` functionality
* `Renderer.redirect` now supports extra arguments which are passed to `Router.linkto` for reverse routing
* new `Renderer.response` method specialized for `ExceptionalResponse`
* consolidated `flash` functionality in dedicated module `Flash` (**breaking**)
* added support for URI segments matching in routes
* refactored the `Route` and `Channel` types
* `ExceptionalResponses` now break the execution flow if thrown from controller hooks
* added `up()` as shortcut for `Genie.startup()` to start the web servers
* internal API consolidation
* added new generic `error-xxx.html` page template
* updated bundled JS and CSS files to newer versions

## v0.15.0 - 2019-08-22

* fixed error in `Genie.newapp()` with `dbsupport = true`

src/REPL.jl
module REPL

using Revise, SHA, Dates, Pkg
using Genie, Genie.Loggers, Genie.Configuration, Genie.Generator, Genie.Tester, Genie.Util, Genie.FileTemplates

const JULIA_PATH = joinpath(Sys.BINDIR, "julia")


"""
    secret_token() :: String

Generates a random secret token to be used for configuring the SECRET_TOKEN const.
"""
function secret_token() :: String
  sha256("$(randn()) $(Dates.now())") |> bytes2hex
end


"""
"""
function copy_fullstack_app(app_path::String = ".") :: Nothing
  cp(joinpath(@__DIR__, "..", "files", "new_app"), app_path)

  nothing
end


"""
"""
function copy_microstack_app(app_path::String = ".") :: Nothing
  mkdir(app_path)

  for f in ["bin", "config", "public", "src",
            ".gitattributes", ".gitignore",
            "bootstrap.jl", "env.jl", "genie.jl", "routes.jl"]
    cp(joinpath(@__DIR__, "..", "files", "new_app", f), joinpath(app_path, f))
  end

  remove_fingerprint_initializer(app_path)

  nothing
end


"""
"""
function copy_db_support(app_path::String = ".") :: Nothing
  cp(joinpath(@__DIR__, "..", "files", "new_app", "db"), joinpath(app_path, "db"))
  cp(joinpath(@__DIR__, "..", "files", "new_app", "config", "initializers", "searchlight.jl"), joinpath(app_path, "config", "initializers", "searchlight.jl"))

  nothing
end


"""
"""
function copy_mvc_support(app_path::String = ".") :: Nothing
  cp(joinpath(@__DIR__, "..", "files", "new_app", "app"), joinpath(app_path, "app"))

  nothing
end


"""
Generates a valid secrets.jl file with a random SECRET_TOKEN.
"""
function write_secrets_file(app_path::String = ".") :: Nothing
  open(joinpath(app_path, Genie.CONFIG_PATH, "secrets.jl"), "w") do f
    write(f, """const SECRET_TOKEN = "$(secret_token())" """)
  end

  nothing
end


"""
"""
function write_app_custom_files(path::String, app_path::String) :: Nothing
  moduleinfo = FileTemplates.appmodule(path)

  open(joinpath(app_path, "src", moduleinfo[1] * ".jl"), "w") do f
    write(f, moduleinfo[2])
  end

  chmod(joinpath(app_path, "bootstrap.jl"), 0o644)

  open(joinpath(app_path, "bootstrap.jl"), "w") do f
    write(f,
    """
      cd(@__DIR__)
      using Pkg
      pkg"activate ."

      function main()
        include(joinpath("src", "$(moduleinfo[1]).jl"))
      end; main()
    """)
  end

  nothing
end


"""
"""
function install_app_dependencies(app_path::String = ".") :: Nothing
  log("Installing app dependencies")
  pkg"activate ."

  pkg"add Genie#master"
  pkg"add JSON"
  pkg"add Millboard"
  pkg"add Revise"

  nothing
end


"""
"""
function autostart_app(autostart::Bool = true) :: Nothing
  if autostart
    log("Starting your brand new Genie app - hang tight!", :info)
    loadapp(".", autostart = autostart)
  else
    log("Your new Genie app is ready!
        Run \njulia> Genie.loadapp() \nto load the app's environment
        and then \njulia> Genie.startup() \nto start the web server on port 8000.")
  end

  nothing
end


"""
"""
function remove_fingerprint_initializer(app_path::String = ".") :: Nothing
  rm(joinpath(app_path, "config", "initializers", "fingerprint.jl"), force = true)

  nothing
end


"""
"""
function remove_searchlight_initializer(app_path::String = ".") :: Nothing
  rm(joinpath(app_path, "config", "initializers", "searchlight.jl"), force = true)

  nothing
end


"""
    newapp(path::String = "."; autostart::Bool = true, fullstack::Bool = false, dbsupport::Bool = false, mvcsupport::Bool = false) :: Nothing

Scaffolds a new Genie app, setting up the file structure indicated by the various arguments.

# Arguments
- `path::String`: the name of the app and the path where to bootstrap it
- `autostart::Bool`: automatically start the app once the file structure is created
- `fullstack::Bool`: the type of app to be bootstrapped. The fullstack app includes MVC structure, DB connection code, and asset pipeline files.
- `dbsupport::Bool`: bootstrap the files needed for DB connection setup via the SearchLight ORM
- `mvcsupport::Bool`: adds the files used for Flax view templates rendering and working with resources

# Examples
```julia-repl
julia> Genie.newapp("MyGenieApp")
2019-08-06 16:54:15:INFO:Main: Done! New app created at MyGenieApp
2019-08-06 16:54:15:DEBUG:Main: Changing active directory to MyGenieApp
2019-08-06 16:54:15:DEBUG:Main: Installing app dependencies
 Resolving package versions...
  Updating `~/Dropbox/Projects/GenieTests/MyGenieApp/Project.toml`
  [c43c736e] + Genie v0.10.1
  Updating `~/Dropbox/Projects/GenieTests/MyGenieApp/Manifest.toml`

2019-08-06 16:54:27:INFO:Main: Starting your brand new Genie app - hang tight!
 _____         _
|   __|___ ___|_|___
|  |  | -_|   | | -_|
|_____|___|_|_|_|___|

┌ Info:
│ Starting Genie in >> DEV << mode
└
[ Info: Logging to file at MyGenieApp/log/dev.log
[ Info: Ready!
2019-08-06 16:54:32:DEBUG:Main: Web Server starting at http://127.0.0.1:8000
2019-08-06 16:54:32:DEBUG:Main: Web Server running at http://127.0.0.1:8000
```
"""
function newapp(path::String = "."; autostart::Bool = true, fullstack::Bool = false, dbsupport::Bool = false, mvcsupport::Bool = false) :: Nothing
  app_path = abspath(path)

  fullstack ? copy_fullstack_app(app_path) : copy_microstack_app(app_path)

  dbsupport ? (fullstack || copy_db_support(app_path)) : remove_searchlight_initializer(app_path)

  mvcsupport && (fullstack || copy_mvc_support(app_path))

  write_secrets_file(app_path)

  write_app_custom_files(path, app_path)

  Sys.iswindows() ? setup_windows_bin_files(app_path) : setup_nix_bin_files(app_path)

  log("Done! New app created at $(abspath(path))", :info)

  log("Changing active directory to $app_path")
  cd(path)

  install_app_dependencies(app_path)

  autostart_app(autostart)

  nothing
end


"""
    loadapp(path::String = "."; autostart::Bool = false) :: Nothing

Loads an existing Genie app from the file system, within the current Julia REPL session.

# Arguments
- `path::String`: the path to the Genie app on the file system.
- `autostart::Bool`: automatically start the app upon loading it.

# Examples
```julia-repl
shell> tree -L 1
.
├── Manifest.toml
├── Project.toml
├── bin
├── bootstrap.jl
├── config
├── env.jl
├── genie.jl
├── log
├── public
├── routes.jl
└── src

5 directories, 6 files

julia> using Genie

julia> Genie.loadapp(".")
 _____         _
|   __|___ ___|_|___
|  |  | -_|   | | -_|
|_____|___|_|_|_|___|

┌ Info:
│ Starting Genie in >> DEV << mode
└
[ Info: Logging to file at MyGenieApp/log/dev.log
```
"""
function loadapp(path::String = "."; autostart::Bool = false) :: Nothing
  Core.eval(Main, Meta.parse("using Revise"))
  Core.eval(Main, Meta.parse("""include(joinpath("$path", "bootstrap.jl"))"""))
  Core.eval(Main, Meta.parse("Revise.revise()"))
  Core.eval(Main, Meta.parse("using Genie"))

  Core.eval(Main.UserApp, Meta.parse("Revise.revise()"))
  Core.eval(Main.UserApp, Meta.parse("$autostart && Genie.startup()"))

  nothing
end


"""
    setup_windows_bin_files(path::String = ".") :: Nothing

Creates the bin/server and bin/repl binaries for Windows
"""
function setup_windows_bin_files(path::String = ".") :: Nothing
  open(joinpath(path, "bin", "repl.bat"), "w") do f
    write(f, "$JULIA_PATH --color=yes --depwarn=no -q -i -- ../bootstrap.jl %*")
  end

  open(joinpath(path, "bin", "server.bat"), "w") do f
    write(f, "$JULIA_PATH --color=yes --depwarn=no -q -i -- ../bootstrap.jl s %*")
  end

  nothing
end


"""
    setup_nix_bin_files(app_path::String = ".") :: Nothing

Creates the bin/server and bin/repl binaries for *nix systems
"""
function setup_nix_bin_files(app_path::String = ".") :: Nothing
  chmod(joinpath(app_path, "bin", "server"), 0o700)
  chmod(joinpath(app_path, "bin", "repl"), 0o700)

  nothing
end

end
